File: cpp/leetcode/369.cpp
class Solution {
public:
    ListNode* plusOne(ListNode* head) {
        if (!head) return head;
        ListNode *rev_head = reverse(head), *cur = rev_head, *pre = cur;
        int carry = 1;
        while (cur) {
            pre = cur;
            int t = cur->val + carry;
            cur->val = t % 10;
            carry = t / 10;
            if (carry == 0) break;
            cur = cur->next;
        }
        if (carry) pre->next = new ListNode(1);
        return reverse(rev_head);
    }
    ListNode* reverse(ListNode *head) {
        if (!head) return head;
        ListNode *dummy = new ListNode(-1), *cur = head;
        dummy->next = head;
        while (cur->next) {
            ListNode *t = cur->next;
            cur->next = t->next;
            t->next = dummy->next;
            dummy->next = t;
        }
        return dummy->next;
    }
};

File: python/leetcode/369.py
class Solution:
    def plusOne(self, head):
        tail = None
        while head:
            head.next, head, tail = tail, head.next, head
        carry = 1
        while tail:
            carry, tail.val = divmod(carry + tail.val, 10)
            if carry and not tail.next:
                tail.next = ListNode(carry)
            tail.next, tail, head = head, tail.next, tail
        return head
